src/app/model/ChessEngine.ts

```diff
@@ -204,4 +217,39 @@ export class ChessEngine implements ChessModel {
   private onKingWasUnchecked(event: KingWasUnchecked): void {
     this.checkedKing = undefined;
   }
+
+  private checkmateHasOccurred(): CheckmateHasOccurred | undefined {
+    if (!this.isKingChecked()) return undefined;
+    if (this.isAnyPossibleMoves()) return undefined;
+
+    const kingPosition = this.kingPosition(this.board, this.currentSide);
+    return {
+      eventType: 'CheckmateHasOccurred',
+      king: new King(this.currentSide),
+      onSquare: kingPosition,
+    };
+  }
+
+  private stalemateHasOccurred(): StalemateHasOccurred | undefined {
```

Comment:
@lamparina obczaj to :)

Reply:
tutaj chyba powinniśmy zagwarantować, że pat może się wydarzyć właśnie tylko wtedy kiedy król nie jest szachowany

Reply:
done